magda-gateway/src/Authenticator.ts

```diff
@@ -43,10 +109,146 @@ export default class Authenticator {
         this.passportSessionMiddleware = passport.session();
     }
 
+    /**
+     * destroy the session. Will do two things:
+     * - Delete session data from session store
+     * - Delete session cookie
+     *
+     * @param {express.Request} req
+     * @param {express.Response} res
+     * @returns {Promise<never>}
+     * @memberof Authenticator
+     */
+    destroySession(
+        req: express.Request,
+        res: express.Response
+    ): Promise<never> {
+        return new Promise((resolve, reject) => {
+            req.session.destroy(err => {
+                if (err) {
+                    // Failed to access session storage to delete session data
+                    reject(err);
+                } else {
+                    resolve();
+                }
+            });
+            const deleteCookieOptions = {
+                ...this.sessionCookieOptions
+            };
+            // --- `clearCookie` works in a way like it will fail to delete cookie if maxAge presents T_T
+            // --- https://github.com/expressjs/express/issues/3856#issuecomment-502397226
+            delete deleteCookieOptions.maxAge;
+            res.clearCookie(DEFAULT_SESSION_COOKIE_NAME, deleteCookieOptions);
```

Comment:
`destroySession` actually will be called for a few different scenarios (including logout).
The destroy session error (if any) is from the session store.
We probably want to mute the error (only log it) as:
- Any leftover session will be auto cleared up by session store's prune process
- Once the cookie is removed from user browser, there would be no functionality impact to users. They will be looked like a user without session
- It will be consistent with how `express-session` handles the situation where the cookie is expired in the user's web browser --- leave the session to the pruning process

@AlexGilleran  Any thoughts on this?

Reply:
I tend to think that logging out means the session is over both on the server side and the client side - once I log out that session is dead.

It's a pretty niche set of circumstances, but say I'm using Magda and somehow I realise that someone's stolen my session cookie (maybe I've been MITM'd somehow). I should be able to sign out, destroy my session and be safe again. But if I log out, the session destruction fails and the cookie is deleted, that session is now going to remain active until it expires and there's nothing I can do about it unless I have access to the database to directly delete it. At least if a failure to destroy the session leaves me logged in I can try again.

Reply:
will create a new commit to:
- make sure logout endpoint become an API (rather simply respond a 302 redirection) and response error if session failed to destroy
- Will also make relevant the frontend logout changes as well